py4j-java/src/main/java/py4j/GatewayServer.java

```diff
@@ -427,6 +427,35 @@ public GatewayServer(Object entryPoint, int port, InetAddress address, int conne
 		this.sSocketFactory = sSocketFactory;
 	}
 
+	/*
+	 * @param gateway gateway instance (or subclass).  Must not be <code>null</code>.
+	 * @param port the host port to usu
+	 * @param address the host address to use
+	 * @param connectTimeout the connect timeout (ms)
+	 * @param readTimeout the read timeout (ms)
+	 * @param customCommands any customCommands to use.  May be <code>null</code>
+	 * @param sSocketFactory socketFactory to use.  Must not be <code>null</code>
+	 */
+	public GatewayServer(Gateway gateway, int port, InetAddress address, int connectTimeout, int readTimeout,
```

Comment:
It is indeed unfortunate and it clearly shows the evolution of the API throughout the years :-) I think I would still prefer to include any new parameters in GatewayServerBuilder. As you say, Gateway already incorporates some of the constructor parameters so if one were to supply a Gateway instance, an entryPoint and a Py4JPythonClient instance, the Gateway instance would take priority and the other two parameters would be ignored (Javadoc could clarify this unfortunate case).

Maybe we can keep two constructors: the one you defined which is the sensible one, and the other that will be used by the builder. At this point, the builder makes more sense for many users considering the large number of parameters (and use cases!).